# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/base/nsFocusManager.cpp
# Commit: 5b44495fa55e
# Full Hash: 5b44495fa55ece2e8b241e83eff82623971f3822
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2023-12-27 20:58:35
# Description:
#   Bug 1872000 - Make `nsFocusManager::GetSelectionLocation` check whether the text frame is nullptr before creating `nsFrameIterator` r=emilio
#   
#   See the comment in the bug, this is not a new crash.  The old factory method
#   did the null-check and returned.  Therefore, before using `MOZ_TRY`, calling
#   a method of `nsFrameIterator` with `nullptr` caused a crash.
# ==============================================================================

diff -r 3c2951a35dcd -r 5b44495fa55e dom/base/nsFocusManager.cpp
--- a/dom/base/nsFocusManager.cpp	Wed Dec 27 07:00:10 2023 +0000
+++ b/dom/base/nsFocusManager.cpp	Wed Dec 27 07:48:25 2023 +0000
@@ -3220,7 +3220,13 @@
       text && text->TextDataLength() == domRange->StartOffset() &&
       domSelection->IsCollapsed()) {
     nsIFrame* startFrame = start->GetPrimaryFrame();
+    // FIXME: If the text node is empty or only collapsible white-spaces next
+    // to a block boundary, it may not have frame, however, we don't know how to
+    // reproduce this.
     MOZ_ASSERT(startFrame);
+    if (MOZ_UNLIKELY(!startFrame)) {
+      return NS_ERROR_FAILURE;
+    }
     // Yes, indeed we were at the end of the last node
     nsIFrame* limiter =
         domSelection && domSelection->GetAncestorLimiter()
